src/compiler/config/validate-paths.ts

```diff
@@ -2,48 +2,67 @@ import { isAbsolute, join } from 'path';
 
 import type * as d from '../../declarations';
 
+/**
+ * The paths validated in this module. These fields can be incorporated into a
+ * {@link d.ValidatedConfig} object.
+ */
+interface ConfigPaths {
+  rootDir: string;
+  srcDir: string;
+  packageJsonFilePath: string;
+  cacheDir: string;
+  srcIndexHtml: string;
+  globalScript?: string;
+  globalStyle?: string;
+  buildLogFilePath?: string;
+}
+
 /**
  * Do logical-level validation (as opposed to type-level validation)
  * for various properties in the user-supplied config which represent
  * filesystem paths.
  *
  * @param config a validated user-supplied configuration
+ * @returns an object holding the validated paths
  */
-export const validatePaths = (config: d.ValidatedConfig) => {
-  if (typeof config.rootDir !== 'string') {
-    config.rootDir = '/';
-  }
+export const validatePaths = (config: d.Config): ConfigPaths => {
+  const rootDir = typeof config.rootDir !== 'string' ? (config.rootDir = '/') : config.rootDir;
```

Comment:
Do we need to mutate `config.rootDir` in this expression? I'm personally just not a fan of assignments _within_ ternary expressions

Reply:
The assignment within the ternary aside, _do_ we need to perform mutation on `config` in this function for `rootDir` and `buildLogFilePath`? Aside from tests, it appears this function is called 2x in `validate-config.ts#validateConfig`:
1. [On entry](https://github.com/ionic-team/stencil/pull/4329/files#diff-53c120bef3205f89229fbe98a5aadfcb51012ead144729ade2fc597c1e9ad5e9R90), to create the `ValidatedConfig`
2. [Far later in the function](https://github.com/ionic-team/stencil/pull/4329/files#diff-53c120bef3205f89229fbe98a5aadfcb51012ead144729ade2fc597c1e9ad5e9R133)

I can see the need for the mutation on `config` if we're concerned that between these two call sites, there's a function that mutates `config`. I haven't dove in deeper here though since @alicewriteswrongs put the time/work in here and could probably set my thinking straight 😆

Reply:
I meant this to be just

```ts
  const rootDir = typeof config.rootDir !== 'string' ? '/' : config.rootDir;
```

and I think I looked at the code for too long and stopped seeing `(config.rootDir = '/')` and started seeing `'/'` instead haha